src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +409,115 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+            using var _ = ArrayBuilder<MatchResult<CompletionItem?>>.GetInstance(out var matchResultsBuilder);
+            var index = 0;
+            var completionHelper = CompletionHelper.GetHelper(document);
+            foreach (var item in completionList.Items)
+            {
+                if (CompletionHelper.TryCreateMatchResult<CompletionItem?>(
+                    completionHelper,
+                    item,
+                    editorCompletionItem: null,
+                    filterText,
+                    completionTrigger.Kind,
+                    GetFilterReason(completionTrigger),
+                    ImmutableArray<string>.Empty,
+                    includeMatchSpans: false,
+                    ref index,
+                    out var matchResult))
+                {
+                    matchResultsBuilder.Add(matchResult);
+                }
+            }
+
+            // Next, we sort the list based on the pattern matching result.
+            var matchedResults = matchResultsBuilder.ToImmutableArray().Sort(MatchResult<CompletionItem?>.SortingComparer);
+
+            // Finally, truncate the list to 1000 items and any preselected items.
+            var count = MaxCompletionListSize;
+            var filteredList = matchedResults
+                .Where(match => count-- > 0 || match.RoslynCompletionItem.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection)
+                .Select(match => match.RoslynCompletionItem)
+                .ToImmutableArray();
+            var newCompletionList = completionList.WithItems(filteredList);
+
+            // Per the LSP spec, the completion list should be marked with isIncomplete = false when further insertions will
+            // not generate any more completion items.  This means that we should be checking if the matchedResults is larger
+            // than the filteredList.  However, the VS client has a bug where they do not properly re-trigger completion
+            // when a character is deleted to go from a complete list back to an incomplete list.
+            // For example, the following scenario.
+            // User types "So" -> server gives subset of items for "So" with isIncomplete = true
+            // User types "m" -> server gives entire set of items for "Som" with isIncomplete = false
+            // User deletes "m" -> client has to remember that "So" results were incomplete and re-request if the user types something else, like "n"
+
+            // Currently the VS client does not remember to re-request, so the completion list only ever shows items from "Som"
+            // so we always set the isIncomplete flag to true when the original list size (computed when no filter text was typed) is too large.
+            // VS bug here - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1335142
+            var isInComplete = completionList.Items.Length > newCompletionList.Items.Length;
```

Comment:
i'd prefer we abide by the spec, and push this on them to fix.  otherwise they may never fix this.

Reply:
I would agree that they need to fix this, however not doing this results in a really not nice experience, if the original list is large I have to delete the entire word and restart to get new completion items.  for this to be useable I think the only option is to do this